feat: Azbatch executor

jakevc opened the pull request:
### Description

Finish remaining issues with the original implementation of the Azure Batch Executor originally tackled by @andreas-wilm in #533. 

- simplified command line options
- batch configuration as a class with sensible defaults, or environment variable overrides.
- made use of existing AzBlob helper where possible.
- updates dockerfile with necessary deps
- container jakevc/snakemake:latest contains these changes for testing

Toyed with the idea of having more configuration from command line options, but this litters the codebase with redundant code (each command line option needs to be repeated a few times in __init__.py, scheduler.py, workflow.py, azure_batch.py). Env Variables seem appropriate with sensible defaults. 

### QC
<!-- Make sure that you can tick the boxes below. -->

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).


### Example

Example invocation:

```bash
export AZ_BLOB_ACCOUNT_URL=*********
export AZ_BATCH_ACCOUNT_KEY=*********

snakemake \
    --jobs 3 \
    -rpf --verbose --default-remote-prefix snaketest \
    --use-conda \
    --default-remote-provider AzBlob \
    --envvars AZ_BLOB_ACCOUNT_URL \
    --az-batch \
    --container-image jakevc/snakemake \
    --az-batch-account-url **********
```

vsmalladi commented:
Is this PR ready to test?

jakevc commented:
yes please give it a try and let me know if you have any issues/recommendations

johanneskoester commented on snakemake/executors/azure_batch.py:
```diff
@@ -0,0 +1,518 @@
+from snakemake.exceptions import WorkflowError
+from snakemake.logging import logger
+from snakemake.common import get_container_image, get_file_hash, async_lock
+from snakemake.remote.AzBlob import AzureStorageHelper
```
Please move this import into the method where it is used. Otherwise, AzBlob will be loaded even if somebody uses nothing from azure, essentially turning the required libs a necessary dependency.

johanneskoester commented on snakemake/__init__.py:
```diff
@@ -2751,6 +2785,33 @@ def parse_cores(cores):
+            sys.exit(1)
+        elif not url_can_parse(args.az_batch_account_url):
+            print(
+                "Error: invalide azure batch account url, please use format: https://{account_name}.{location}.batch.azure.com."
```
What is the rationale for representing this as url instead of expecting a location and an account flag?

jakevc replied:
Since the full batch_account_url is used as an argument to BatchServiceClient in line 165, and the batch_account_name is used for the SharedKeyCredentials in line 162, but the location is never used. 

It's a tradeoff I guess. Having it the other way would mean increasing the probability of user error when constructing the batch_account_url, if either the location or the name is mis-specified. This way someone only has to enter the correct value once. Also since location is not used, that flag would be an additional 5ish lines of code throughout the argument parsing logic that really is only used to construct the full batch_account_url and nothing else..

johanneskoester reviewed (changes requested):
Awesome work! Very clean code! Just a few comments below.

And then: I've contacted MS about credits, hopefully we can add a testcase before merging.

johanneskoester commented:
Hey Jake. As Adreas Wilm is already an author of the Snakemake paper because of his contribution to Azure Batch support, it makes total sense to make you a co-author as well in the upcoming revision. Do you agree? If so, please give me your affiliation here.

jakevc commented:
> Hey Jake. As Adreas Wilm is already an author of the Snakemake paper because of his contribution to Azure Batch support, it makes total sense to make you a co-author as well in the upcoming revision. Do you agree? If so, please give me your affiliation here.

Yes that sounds good, thank you!

Please include these two affiliations:
- Earle A. Chiles Research Institute, Portland, Oregon.
- Providence Cancer Institute, Portland, Oregon.

vsmalladi commented on docs/executor_tutorial/azure_batch.rst:
```diff
@@ -0,0 +1,255 @@
+
+.. code:: console
+
+   stgkey=$(az storage account keys list -g $resgroup -n $storageacct | head -n1 | cut -f 3)
```
Should change variable for storage accoutn name to `stgacct` from `storageacct`
```
stgkey=$(az storage account keys list -g $resgroup -n $stgacct | head -n1 | cut -f 3)
```

vsmalladi commented on docs/executor_tutorial/azure_batch.rst:
```diff
@@ -0,0 +1,255 @@
+
+.. code:: console
+
+    az batch account keys --resource-group $resgroup --name $accountname --key-name primary
```
Stoge batch account keys

```
az_batch_account_key=$(az batch account keys list --resource-group $resgroup --name $accountname -o tsv | head -n1 | cut -f2)
```

vsmalladi commented on docs/executor_tutorial/azure_batch.rst:
```diff
@@ -0,0 +1,255 @@
+
+.. code:: console
+
+     export AZ_BLOB_ACCOUNT_URL='${storage_account_url_with_sas}'
```
Do you want users to generate this in CLI or by Portal?

jakevc replied:
CLI would be nice but I can't seem to find the command to generate an SAS url for a storage account that produces the full URI.

vsmalladi replied:
export sastoken=$(az storage account generate-sas \
    --account-name $stgacct \
    --permissions acdlrw \
    --services bf \
    --resource-types co \
    --expiry $expiry_date)

export sas_url=$(az storage account show-connection-string \
    -g $resgroup \
    -n $stgacct \
    --protocol https \
    -o tsv | cut -f5,9 -d ';' | cut -f 2 -d '=')

 export storage_account_url_with_sas='${sas_url}${sastoken}'

Something like this. can be made better.

vsmalladi commented on docs/executor_tutorial/azure_batch.rst:
```diff
@@ -0,0 +1,255 @@
+
+Below we will task Snakemake to install software on the fly with conda.
+For this we need a Snakefile with corresponding conda environment
+yaml files. You can download the package containing all those files `here <workflow/snakedir.zip>`__.
```
Where is this file, link doesn't work

vsmalladi replied:
Should we include this in the repo?

vsmalladi reviewed (commented):
When I ran through the turorial I got the following error:

```
/Users/venkatmalladi/opt/anaconda3/envs/snakemake/lib/python3.11/site-packages/snakemake/executors/azure_batch.py:352: RuntimeWarning: coroutine 'sleep' was never awaited
  sleep()
```

vsmalladi commented:
Overall this is great. Does this scale to be a new node for each task or all the tasks running on a single node?

Just curious since conda is still used vs docker

vsmalladi commented:
Got this error on mac os: 

```
The Python 3 package 'azure-storage-blob' need to be installed to use Azure Storage remote() file functionality. No module named 'azure'
```

Verified that azure-storage-blob was installed using pip. Any ideas?

jakevc commented:
> When I ran through the tutorial I got the following error:
> 
> ```
> /Users/venkatmalladi/opt/anaconda3/envs/snakemake/lib/python3.11/site-packages/snakemake/executors/azure_batch.py:352: RuntimeWarning: coroutine 'sleep' was never awaited
>   sleep()
> ```
> 
> No issue once i exported AZ_BATCH_ACCOUNT_KEY

fixed this

jakevc commented:
> Overall this is great. Does this scale to be a new node for each task or all the tasks running on a single node?
> 
> Just curious since conda is still used vs docker

Currently the pool autoscaling is not enabled, so it will distribute tasks across one node (however big it is). In the future I would like to include autoscaling, but need to do some testing to configure the most appropriate autoscaling formula, discussed here: https://learn.microsoft.com/en-us/azure/batch/batch-automatic-scaling

vsmalladi commented:
@jakevc I tried to run this again.

Based on the tutorial the base snakemake image might not have the azure plugins cause I get this error:

`The Python 3 package 'azure-storage-blob' need to be installed to use Azure Storage remote() file functionality. No module named 'azure'`

Thoughts?

jakevc commented:
> @jakevc I tried to run this again.
> 
> Based on the tutorial the base snakemake image might not have the azure plugins cause I get this error:
> 
> `The Python 3 package 'azure-storage-blob' need to be installed to use Azure Storage remote() file functionality. No module named 'azure'`
> 
> Thoughts?

I will try this out in a fresh environment and see if I can reproduce

jakevc commented:
> scale to be a new node for each task or all the tasks running on a single node?
> 
> Just curious since conda is still used vs docker

Was able to implement basic autoscaling, and enabled some other configuration parameters that enable task distribution on dedicated ( non autoscaling ) nodes as well.

vsmalladi commented:
> > scale to be a new node for each task or all the tasks running on a single node?
> > Just curious since conda is still used vs docker
> 
> Was able to implement basic autoscaling, and enabled some other configuration parameters that enable task distribution on dedicated ( non autoscaling ) nodes as well.

Oh that is cool. Looking forward to testing that out.

jakevc commented:
> @jakevc I tried to run this again.
> 
> Based on the tutorial the base snakemake image might not have the azure plugins cause I get this error:
> 
> `The Python 3 package 'azure-storage-blob' need to be installed to use Azure Storage remote() file functionality. No module named 'azure'`
> 
> Thoughts?

Okay the method I am using to create an environment and test this branch is the following, and you should be able to reproduce if you follow these instructions.

Fist checkout my fork branch `azbatch-executor` and using either mamba or conda to create the testing environment. Mamba solves the environment a lot faster.  From the root of the snakemake repo:

```
mamba env create -f test-environment.yml -n snakemake-testing
conda activate snakemake-testing
pip install -e ".[azure]"
```

You should then be able to run the example.

vsmalladi commented:
Where should the output files be stored once the pipeline is done?

jakevc commented:
In the storage container/prefix specified at runtime by `--default-remote-prefix`

vsmalladi commented:
Okay ran the update version and am getting an error it can't find the output files:

```
rule bwa_map:
    input: snakemake-tutorial/data/genome.fa, snakemake-tutorial/data/genome.fa.amb, snakemake-tutorial/data/genome.fa.ann, snakemake-tutorial/data/genome.fa.bwt, snakemake-tutorial/data/genome.fa.fai, snakemake-tutorial/data/genome.fa.pac, snakemake-tutorial/data/genome.fa.sa, snakemake-tutorial/data/samples/B.fastq
    output: snakemake-tutorial/mapped_reads/B.bam
    log: snakemake-tutorial/logs/mapped_reads/B.log
    jobid: 5
    reason: Missing output files: snakemake-tutorial/mapped_reads/B.bam
    wildcards: sample=B
    threads: 2
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=<TBD>

{ bwa mem -t 2 snakemake-tutorial/data/genome.fa snakemake-tutorial/data/samples/B.fastq | samtools view -Sb - > snakemake-tutorial/mapped_reads/B.bam; } 2>snakemake-tutorial/logs/mapped_reads/B.log
```

jakevc commented:
The tutorial expects the input files: `snakemake-tutorial/data/genome.fa, snakemake-tutorial/data/genome.fa.amb, snakemake-tutorial/data/genome.fa.ann, snakemake-tutorial/data/genome.fa.bwt, snakemake-tutorial/data/genome.fa.fai, snakemake-tutorial/data/genome.fa.pac, snakemake-tutorial/data/genome.fa.sa, snakemake-tutorial/data/samples/B.fastq` to exist in the storage account you are using, with `snakemake-tutorial` as the storage container.

jakevc commented:
Can you post/attach the full output log or error?

vsmalladi commented:
[2022-12-22T192940.156414.snakemake.log](https://github.com/snakemake/snakemake/files/10292068/2022-12-22T192940.156414.snakemake.log)

jakevc commented:
Can you confirm that you have all input files where they are suppose to be. `snakemake-tutorial/data/genome.fa, snakemake-tutorial/data/genome.fa.amb, snakemake-tutorial/data/genome.fa.ann, snakemake-tutorial/data/genome.fa.bwt, snakemake-tutorial/data/genome.fa.fai, snakemake-tutorial/data/genome.fa.pac, snakemake-tutorial/data/genome.fa.sa, snakemake-tutorial/data/samples/B.fastq`. Those files should all exist in the storage account you are using. With "snakemake-tutorial" as the container name.

jakevc commented:
And there should also be `snakemake-tutorial/data/samples/A.fastq, snakemake-tutorial/data/samples/B.fastq, snakemake-tutorial/data/samples/C.fastq`

vsmalladi commented:
Yep all present

vsmalladi commented:
okay it looks like it is not pulling in the data from the storage account and I think its trying to get the files from git:

```
Unable to retrieve additional files from git. This is not a git repository.
```

vsmalladi commented:
Confirmed this works on azure. I think we can merge this in when ready.

vsmalladi commented:
@jakevc @johanneskoester Any blockers on this PR?

jakevc commented:
> @jakevc @johanneskoester Any blockers on this PR?

All good on my end

johanneskoester commented:
Awesome, thanks a lot! Indeed we have been granted credits by MS! I will try to include the tokens into the secrets ASAP.

johanneskoester commented:
Subsciption to azure is set up, and I have added the credentials as AZ_BATCH_KEY and AZ_STORAGE_KEY in Github Actions. Can you add a test case here?

jakevc commented:
> Subsciption to azure is set up, and I have added the credentials as AZ_BATCH_KEY and AZ_STORAGE_KEY in Github Actions. Can you add a test case here?

Okay I have added a test case. It requires more secrets, please add the remaining secrets referenced in the workflow:

```
  export AZ_BLOB_ACCOUNT_URL=${{ secrets.AZ_STORAGE_ACCOUNT_URL }}
  export AZ_BLOB_AZ_BLOB_CREDENTIAL=${{ secrets.AZ_STORAGE_KEY }}
  export AZ_BATCH_ACCOUNT_URL=${{ secrets.AZ_BATCH_ACCOUNT_URL }}
  export AZ_BATCH_ACCOUNT_KEY=${{ secrets.AZ_BATCH_KEY }}
  pytest -s -v -x tests/test_azure_batch_executor.py
```

johanneskoester commented on tests/test_azure_batch_executor.py:
```diff
@@ -0,0 +1,20 @@
+def test_az_batch_executor():
+    bau = os.getenv("AZ_BATCH_ACCOUNT_URL")
+    prefix = os.getenv("AZ_BLOB_PREFIX")
+    wdir = dpath("test_azure_batch")
```
I think the test folder still needs to be added.

jakevc replied:
ug yeah git really wants to ignore that whole directory for some reason

johanneskoester commented on tests/test_azure_batch_executor.py:
```diff
@@ -0,0 +1,20 @@
+    run(
+        path=wdir,
+        default_remote_prefix=prefix,
+        container_image="jakevc/snakemake",
```
Why is a special container image needed?

vsmalladi replied:
It shouldn't I think we need to back into base snakemake image

johanneskoester commented on tests/test_azure_batch_executor.py:
```diff
@@ -0,0 +1,20 @@
+
+
+def test_az_batch_executor():
+    bau = os.getenv("AZ_BATCH_ACCOUNT_URL")
```
I am unsure how this should look like and where I will find it. Can you also put this info in the section of the docs? For now, I've set it to `https://snakemakebatch.eastus.batch.azure.com`.

jakevc commented:
@johanneskoester please make sure the secret AZ_STORAGE_ACCOUNT_URL is formatted as `https://accountname.blob.core.windows.net`

johanneskoester commented:
> @johanneskoester please make sure the secret AZ_STORAGE_ACCOUNT_URL is formatted as `https://accountname.blob.core.windows.net`

Ok, done!

vsmalladi commented:
@johanneskoester is this ready to merge or do we need more testing?

johanneskoester commented:
@jakevc the test fails, can you please have a look?

jakevc commented:
@johanneskoester  It doesn't look like the test action is picking up the secret values correctly. Can you double check they are present and corresponding to below: 

```
          export AZ_BLOB_PREFIX=${{ secrets.AZ_BLOB_PREFIX }}
          export AZ_BLOB_ACCOUNT_URL=${{ secrets.AZ_STORAGE_ACCOUNT_URL }}
          export AZ_BLOB_CREDENTIAL=${{ secrets.AZ_STORAGE_KEY }}
          export AZ_BATCH_ACCOUNT_URL=${{ secrets.AZ_BATCH_ACCOUNT_URL }}
          export AZ_BATCH_ACCOUNT_KEY=${{ secrets.AZ_BATCH_KEY }}
```

jakevc commented:
This is the format of the secrets I am using when I run the test locally and it passes:

AZ_BLOB_PREFIX=snaketest
AZ_BLOB_ACCOUNT_URL=https://***********.blob.core.windows.net/
AZ_BATCH_ACCOUNT_URL=https://**********.westus2.batch.azure.com
AZ_BLOB_CREDENTIAL=$SAS
AZ_BATCH_ACCOUNT_KEY=$KEY

johanneskoester commented:
No clue, I just checked it again. The URL for the storage account seems follow the same pattern as you suggest.

johanneskoester commented:
Maybe you can add some parsing of those env vars to the remote provider and the executor, which gives a proper error message if something is invalid (without printing any secrets of course).

vsmalladi commented:
I wonder if there is some special characters that aren't escaped correctly. 

@jakevc can you try dummy variables and print those to stdout and see if there is something odd with the variables.

vsmalladi commented:
@jakevc  can you do this

          AZ_BLOB_PREFIX: '$${{ secrets.AZ_BLOB_PREFIX }}'
          AZ_BLOB_ACCOUNT_URL: '${{ secrets.AZ_STORAGE_ACCOUNT_URL }}'
          AZ_BLOB_CREDENTIAL: '${{ secrets.AZ_STORAGE_KEY }}'
          AZ_BATCH_ACCOUNT_URL: '${{ secrets.AZ_BATCH_ACCOUNT_URL }}'
          AZ_BATCH_ACCOUNT_KEY: '${{ secrets.AZ_BATCH_KEY }}'

jakevc commented:
Face palm: of course this is the issue. The funny thing is on my development environment zsh doesn't even let me set the blob credential without quoting it.

jakevc commented:
or not...

jakevc commented:
Okay I'm not crazy.. I slightly more specific error and now we are failing with:

```
raise ValueError("Blob Account URL is None or empty string")
[220](https://github.com/snakemake/snakemake/actions/runs/4848828641/jobs/8640274478#step:20:221)
ValueError: Blob Account URL is None or empty string
```

johanneskoester commented:
Understood the problem now! The issue is that as the PR comes from a fork, it does not have access to secrets.

johanneskoester commented:
I've merged your changes in a snakemake-owned branch and created a PR for merging into main here: https://github.com/snakemake/snakemake/pull/2246
Let us go on from there. As you are anyway already part of the Snakemake org, you should in the future work directly in our repo in order to ensure that your PRs always have the secrets.

vsmalladi commented on snakemake/remote/AzBlob.py:
```diff
@@ -182,11 +182,35 @@ def __init__(self, *args, **kwargs):
+        bool: True if the environment variable is a valid Azure Storage Account URL, False otherwise.
+        """
+        url_pattern = re.compile(
+            r"^https:\/\/[a-z0-9]+(\.[a-z0-9]+)*\.blob\.core\.windows\.net\/?$"
```
@jakevc this url pattern fails when the sas token is added  should be

`r"^https:\/\/[a-z0-9]+(\.[a-z0-9]+)*\.blob\.core\.windows\.net\/(\?.+)?$"`
